lasagne/layers/transform.py

```diff
@@ -0,0 +1,207 @@
+
+__all__ = [
+    "TransformerLayer",
+]
+
+
+class TransformerLayer(MergeLayer):
+    """
+    Spatial transformer layer
+
+    The layer applies an affine transformation on the input. The affine
+    transformation is parameterized with six learned parameters [1]_.
+    The output is interpolated with a bilinear transformation.
+
+    Parameters
+    ----------
+    input : a :class:`Layer` instance
+        The input where the affine transformation is applied. This should
+        have convolution format, i.e (num_batch, channels, height, width).
+    localization_network: : a :class:`Layer` instance
+        The network that calculates the parameters of the affine
+        transformation. See the example for how to initialize to the identity
+        transform.
+    downsample_factor : float
+        Determines the size of the output image. A value of 1 will keep the
+        original size of the input. Values larger than 1 will down sample the
+        input. Values below 1 will up sample the input.
+
+    References
+    ----------
+    .. [1]  Spatial Transformer Networks
+            Max Jaderberg, Karen Simonyan, Andrew Zisserman, Koray Kavukcuoglu
+            Submitted on 5 Jun 2015
+
+    Examples
+    --------
+    Here we set up the layer to initially do the identity transform, similarly
+    to [1]_. Note that you will want to use a localization with linear output.
+    If the output from the localization networks is [t1, t2, t3, t4, t5, t6]
+    then t1 and t5 determines zoom, t2 and t4 determines skewness, and t3 and
+    t6 move the center position.
+
+    >>> import numpy as np
+    >>> import lasagne
+    >>> b = np.zeros((2, 3), dtype='float32')
+    >>> b[0, 0] = 1
+    >>> b[1, 1] = 1
+    >>> b = b.flatten()  # identity transform
+    >>> W = lasagne.init.Constant(0.0)
+    >>> l_in = lasagne.layers.InputLayer((None, 3, 28, 28))
+    >>> l_loc = lasagne.layers.DenseLayer(l_in, num_units=6, W=W, b=b,
+    ... nonlinearity=None)
+    >>> l_trans = lasagne.layers.TransformerLayer(l_in, l_loc)
+    """
+    def __init__(
+            self, input, localization_network, downsample_factor=1, **kwargs):
+        super(TransformerLayer, self).__init__(
+            [input, localization_network], **kwargs)
+        self.downsample_factor = downsample_factor
+
+        input_shp, loc_shp = self.input_shapes
+
+        if loc_shp[-1] != 6 or len(loc_shp) != 2:
+            raise ValueError("The localization network must have "
+                             "output shape [num_batch, 6]")
+        if len(input_shp) != 4:
+            raise ValueError('The input network must have a 4-dimensional '
+                             'output shape:(batch_size, num_input_channels, '
+                             'input_rows, input_columns)')
+
+    def get_output_shape_for(self, input_shapes):
+        shp = input_shapes[0]
+        dsf = self.downsample_factor
+        return shp[:2] + tuple(int(s//dsf) for s in shp[2:])
```

Comment:
And maybe we should't use integer division if we allow floats for downsample factors?